vivarium/core/process.py

```diff
@@ -280,6 +268,41 @@ def get_parameters(self):
             process_id: process.parameters
             for process_id, process in processes.items()}
 
+    # def merge(self, processes={}, topology={}, schema_override={}):
+    #     for name, process in processes.items():
+    #         assert isinstance(process, Process)
+    #
+    #     self.merge_processes = deep_merge_check(self.merge_processes, processes)
+    #     self.merge_topology = deep_merge(self.merge_topology, topology)
+    #     self.schema_override = deep_merge(self.schema_override, schema_override)
+
+
+class Composite(Generator):
+    """Composite parent class
+
+    All :term:`composite` classes must inherit from this class.
```

Comment:
Including `ToyComposite` and `InjectedGlcPhosphorylation`.

It looks like we should also add abstract classes `Compartment`, `Port`, and `Agent`.

Reply:
A `Port` is an declared by the `Process` `ports_schema()` method. `Process` does also have a `self.ports()` method for getting the port names from `ports_schema()`. I'm not sure what `Compartment` and `Agent` would gain us -- they are just different applications of `Composite`. Would it be useful to have them even if they are identical at present?

Reply:
Other than the objects that define interfaces, there's no clear answer on which ones to make classes. Additional definitions entail some mental overhead, so it's wise to question the utility.

Some benefits even if the classes would be identical at present:

* Navigation: grep the code to find the Agent classes, navigate the class structure in PyCharm, and trace it at runtime. A grep for classes defined from `Generator` found classes with obvious names like `ToyCompartment`; `SplitPort` and `MergePort` which gave me the misimpression that they're Ports; `InjectedGlcPhosphorylation` which I assume is a `Composite` because it's in the composites directory; and `BB`, `PoQo`, and `MultiUnits` which I didn't look into.
* Hook: Each class is a place to add future behavior to all of its subclasses, e.g. a `Compartment` `__str__` method or an `Agent` debug feature, or to add to the interface protocol along with a default implementation. This helps more in a library with external users than in an application where we can change something and update all the relevant code.
* Documentation: Generate class hierarchy diagrams. This example is limited but I only had to right-click a class and pick Diagrams > Show Diagram Popup:
![image](https://user-images.githubusercontent.com/1043120/102860698-36258d80-43e3-11eb-8c23-98926df9b27d.png)

That's still not a clear tradeoff.